datumaro/components/extractor.py

```diff
@@ -170,22 +178,97 @@ def get(self, id, subset=None):
                 return item
         return None
 
+T = TypeVar('T')
+
+class _ImportFail(DatumaroError):
+    pass
+
+class ImportErrorPolicy:
+    def report_item_error(self, error: Exception, *,
+            item_id: Tuple[str, str]):
+        """
+        Allows to report a problem with a dataset item.
+        If this function returns, the extractor must skip the item.
+        """
+
+        if not isinstance(error, _ImportFail):
+            ie = ItemImportError(item_id)
+            ie.__cause__ = error
+            return self._handle_item_error(ie)
+        else:
+            raise error
+
+    def report_annotation_error(self, error: Exception, *,
+            item_id: Tuple[str, str]):
+        """
+        Allows to report a problem with a dataset item annotation.
+        If this function returns, the extractor must skip the item.
+        """
+
+        if not isinstance(error, _ImportFail):
+            ie = AnnotationImportError(item_id)
+            ie.__cause__ = error
+            return self._handle_annotation_error(ie)
+        else:
+            raise error
+
+    def _handle_item_error(self, error: ItemImportError):
+        """This function must either call fail() or return."""
+        raise NotImplementedError("Must be defined in the subclass")
+
+    def _handle_annotation_error(self, error: AnnotationImportError):
+        """This function must either call fail() or return."""
+        raise NotImplementedError("Must be defined in the subclass")
+
+    def fail(self, error: Exception) -> NoReturn:
+        raise _ImportFail from error
```

Comment:
Why is `_ImportFail` necessary? What if this function just raised `error`?

Reply:
It is used to indicate exit from importing process. Other exceptions can be captured by other methods of this class.

Reply:
> Other exceptions can be captured by other methods of this class.

Isn't that a good thing, though? Imagine that you want to ignore items that failed to be imported for any reason (including bad annotations). You could write:

```
    def _handle_item_error(self, error: ItemImportError):
        pass

    def _handle_annotation_error(self, error: AnnotationImportError):
        ???
```

But if `_handle_annotation_error` calls `fail`, then the whole import fails; while if it does nothing, the _annotation_ is ignored, and the item is still used. It seems like it would be better if `_handle_item_error` got a shot at handling the error raised by `_handle_annotation_error`.

Reply:
This is a good question. The behavior is going to be quite inconsistent, because we can have any of the following cases in different formats:
```
[try] load item()
    [try] load annotations()

[try] load item()
[try] load annotations()

[try] load item and annotations()
```
And if we have eg. handle_ann -> fail, handle_item -> skip, then in the second case we'll fail, then in the first and the third we will skip. If we don't capture errors, we can also do any of them. If we assume the format is "right" and allows this, then yes. Probably, we can just throw `error`, but then also allow to capture the error in the `report_item_error`. We still need to wrap errors in dedicated classes to indicate their kinds.